lib/boundkeypair/bound_keypair.go

```diff
@@ -0,0 +1,150 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package boundkeypair
+
+import (
+	"crypto"
+	"crypto/subtle"
+	"time"
+
+	"github.com/go-jose/go-jose/v3/jwt"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// challengeExpiration is the TTL for a bound keypair challenge document
+	// once generated by the server, used to calculate the JWT `exp` field.
+	challengeExpiration time.Duration = time.Minute
+
+	// challengeNotBeforeOffset is the offset applied to the `nbf` field to
+	// allow for a small amount of clock drift.
+	challengeNotBeforeOffset = -10 * time.Second
+)
+
+// ChallengeDocument is a bound keypair challenge document. These documents are
+// sent in JSON form to clients attempting to authenticate, and are expected to
+// be sent back signed with a known public key.
```

Comment:
Sorry to nitpick! Feel free to re-word too.

```suggestion
// be sent back signed with the private counterpart of a known public key.
```